Changes to allow customize deploy path for self-hosting

celadevra opened the pull request:
Hi, first please allow me to thank you for bringing ruhoh.

When I deploy ruhoh to my VPS on Linode, I encountered a problem: my blog sits at http://idenizen.net/blog, for testing purposes, I decided to put ruhoh compiled files under http://idenizen.net/ruhoh.  When I browse the url, I find that all asset links and permalinks are all pointing to /, e.g. http://idenizen.net/assets/... .  It seems that although you have `production_url` in `config.yml`, the urls generated doesn't respect this variable.  This can be circumvented by using a virtual host, but currently that's not my idea about my website.

Thus I changed the files a bit, mostly under `lib/ruhoh`, to make ruhoh recognize the path it is in on the server.  Moreover, I set up another variable in `config.yml` in my ruhoh blog, namely `dev_url`.  Both `production_url` and `dev_url` defaults to `/`, and when `config.env` is set to `production`, `production_url` overrides `dev_url`, otherwise `dev_url` is used when `config.env` is set to anything else.  This enables one to preview changes to the blog locally, and by switching to `production`, one can easily upload the results to the production server.  Please see celadevra/celadevra.ruhoh.com@bb49362 for my config.  Also, I find that after these changes, the `div.navbar>div.navbar-inner>a.brand` doesn't give the correct root url, therefore I also changed the layout under the theme a bit, please see the above blob as well.

I tested these changes with my own blog, both locally and on my VPS.  As you may already find out, I am not too familiar with Ruby (I learn some of it from your code :), so please indulge me for bugs.  Also, if ruhoh as of now actually supports deploying to places like somehost.com/your/path, please let me know how.  Much appreciated.

celadevra committed:
Now the urls are relative

and ruhoh blog can be deployed under a subdirectory

celadevra committed:
Partly solved the url problem

celadevra committed:
Fix urls for pages

celadevra committed:
The change of these files allow custom root url

The user can set production_url and dev_url variables in config.yml, if
they are not present, both are default to '/'.  Another variable env
determines whether production_url or dev_url is used as root url.  If
env == 'production', production_url is used, or else dev_url or simply
'/'.  The production_url can be a subdirectory on your www root
directory, or a virtual host, etc.

celadevra committed:
Fix redundant sub()

celadevra committed:
Fix url problems in dev environment

celadevra committed:
Refine asset url when production_url == '/'

Before this commit, when production_url == '' or '/', in the payload the
link to asset will lose the leading '/', which is not right.  By
changing the regex sub used for generating url this is corrected.

celadevra commented:
My bad.  Turns out 4da8463 didn't completely solve the problem.  It gives current page's url in the nav bar, so in terms of behaviour, things didn't change much.

I also find that `production_url` and `dev_url` doesn't default to `/` as I claimed.  I corrected them, wondering if there are more elegant solutions.  My codes looks chaotic ;).

I updated the commit range for the changes, please see celadevra/celadevra.ruhoh.com@69cf7e7 for the changes I made to `layout/default.html`.

plusjade commented:
@celadevra thanks for your work! Contributions are always appreciated. I am starting a new job next week and will be busy this weekend but will give this a look and offer feedback as soon as I can. Thanks for your patience!

celadevra committed:
Fix rss links and limit number of posts in rss

My customroot fix break the url handling in rss.rb.  The generated link
looks like http://somesite.nethttp://somesite.net/archive... so I have
to remove the first docroot string with regexp.

I have over 600 posts now, which are all turned into rss by default,
which causes a xml over 5 MB that is cumbersome to handle.  Now the
number of posts in rss is limited to 20.  Maybe I can change it to a
number set in config.

celadevra committed:
Now the number of posts in rss is variable

in config.yml in your blog

celadevra commented:
My modification of the role of `production_url` breaks the url in the RSS.  The url is used twice so your link looks like http://yoursite.comhttp://yoursite.com/2012... .  As a result I modified rss.rb a bit to generate the right url for self-hosting.  But not sure if it still gives right url on ruhoh.com.  

Meanwhile, as I migrated my 600-some posts to ruhoh, the rss.xml file generated is too big for the server to handle.  I limited the number of posts in the xml file to a value customizable by the user (see celadevra/celadevra.ruhoh.com@bff33d1), say 20.  The new rss.rb is at 12589f04b3.  But I'm not sure if there would be problem if you have less posts than 20, please test if you are starting afresh with ruhoh.  Other static blog generators such as Octopress seems to also limit the number of posts in RSS to 20 or 10, so I think this change make sense.

celadevra committed:
Minor tweak so Home has right link in Pages page

celadevra committed:
Revert link generation method in rss.rb

Actually I don't have to remove docroot from the url in rss.

Previous behaviour may be due to large rss file and is actually an
error.

celadevra committed:
Fix preview in production mode

To let route.rb find correct id for url generation

plusjade commented:
@celadevra I finally addressed the url base_path issue you are trying to solve.
the feature is finished and available in this pull request: https://github.com/ruhoh/ruhoh.rb/pull/78.

The code you have here touches too many different parts of the app and uses multiple configuration variables.
I definitely took inspiration from your work -- let me know if this solves your use case.
